### PR TITLE
Use event watcher for pending transaction pool

### PR DESCRIPTION
**Pull request by kosecki123**
* Refactor network gas price fetching
* Event based pending pool

### CODE REVIEW

File: src/TxPool/TxPool.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+  public running() {
+    return this.pool.size > 0 || !!this.subs.pending || !!this.subs.latest;
+  }
+
+  public async start() {
+    if (this.running()) {
+      await this.stop();
+    }
+    await this.watchPending();
+    if (this.running()) {
```
Comment (adibas03) [inline src/TxPool/TxPool.ts]: This code block is a repetition of `await this.stop()`

**Review by josipbagaric — APPROVED** [review]

**Review by lsaether — APPROVED** [review]

File: src/TxPool/TxPool.ts [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+  constructor(web3: any, util: W3Util, logger: ILogger) {
+    this.web3 = web3;
+    this.logger = logger;
+    this.util = util;
+    this.pool = new Map<string, ITxPoolTxDetails>();
+    this.txPoolProcessor = new TxPoolProcessor(this.util, this.logger);
+  }
+
+  public running() {
+    return !!this.subs[EXECUTED_EVENT] && !!this.subs[CLAIMED_EVENT];
```
Comment (Kuzirashi) [inline src/TxPool/TxPool.ts]: You could have one operation less if you would do:
`!!(this.subs[EXECUTED_EVENT] && this.subs[CLAIMED_EVENT]);`

File: src/TxPool/TxPool.ts [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    if (this.subs.mined) {
+      clearInterval(this.subs.mined);
+    }
+    this.subs = {};
+    this.logger.debug('TxPool STOPPED');
+  }
+
+  private async stopTopic(topic: string) {
+    if (this.subs[topic]) {
+      await this.util.stopFilter(this.subs[topic]).catch(e => {
```
Comment (Kuzirashi) [inline src/TxPool/TxPool.ts]: You either should use `await/async/try/catch` or `.then/.catch`. Here it's mixed, could you switch from using `.catch` to putting `await this.util.stopFilter()` instide `try {} catch () {}` block?

File: src/TxPool/TxPool.ts [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+      this.txPoolProcessor.process(err, res, this.pool)
+    );
+  }
+
+  private clearMined() {
+    this.subs.mined = setInterval(() => {
+      const now = new Date().getTime();
+      this.pool.forEach((value, key) => {
+        if (now - value.timestamp > TIME_IN_POOL) {
+          this.pool.delete(key);
```
Comment (Kuzirashi) [inline src/TxPool/TxPool.ts]: Is this secure to delete from the same array that you iterate over? Would it be safer to run a foreach loop on copy of the array?

**Review by e00dan — APPROVED** [review]

**Pull request merged**